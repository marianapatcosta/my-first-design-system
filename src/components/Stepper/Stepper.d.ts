import React, { CSSProperties, MouseEventHandler, ReactNode } from 'react';
import { ORIENTATION } from '../../constants';
declare type orientationTypes = ORIENTATION;
export interface StepperProps {
    /**
     * element id
     */
    id?: string;
    /**
     * additional css class
     */
    className?: string;
    /**
     * additional style
     */
    style?: CSSProperties;
    /**
     * stepper's orientation
     */
    orientation?: orientationTypes;
    /**
     * label of stepper's button in the last step
     */
    buttonLabel?: string;
    /**
     * Array of objects defining the metadata for each stepper's steps
     */
    stepsMetadata: Array<{
        content: ReactNode;
        isValid?: boolean;
    }>;
    /**
     * Click handler to call on finish the fulfill/visualization of all steps
     */
    onSubmit: MouseEventHandler<HTMLButtonElement>;
}
export declare const Stepper: React.FC<StepperProps>;
export {};
