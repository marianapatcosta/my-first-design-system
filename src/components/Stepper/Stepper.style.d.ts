/// <reference types="react" />
export declare const StyledStepper: import("styled-components").StyledComponent<"div", any, {
    isVertical?: boolean | undefined;
}, never>;
export declare const StyledStepperIndicator: import("styled-components").StyledComponent<"ul", any, {
    isVertical?: boolean | undefined;
}, never>;
export declare const StyledStepperIndicatorItem: import("styled-components").StyledComponent<"li", any, {
    isFilled?: boolean | undefined;
}, never>;
export declare const StyledStepperProgressBar: import("styled-components").StyledComponent<"progress", any, {
    isVertical?: boolean | undefined;
}, never>;
export declare const StyledStepperContainer: import("styled-components").StyledComponent<"div", any, {
    isVertical?: boolean | undefined;
}, never>;
export declare const StyledStepperContent: import("styled-components").StyledComponent<"div", any, {
    isFilled?: boolean | undefined;
}, never>;
export declare const StyledStepperFooter: import("styled-components").StyledComponent<"footer", any, {
    isFilled?: boolean | undefined;
}, never>;
export declare const StyledStepperButton: import("styled-components").StyledComponent<import("react").ForwardRefExoticComponent<import("../Button/Button").ButtonProps & import("react").RefAttributes<HTMLButtonElement>>, any, {
    size: "small";
}, "size">;
export declare const StyledStepperButtonRight: import("styled-components").StyledComponent<import("react").ForwardRefExoticComponent<import("../Button/Button").ButtonProps & import("react").RefAttributes<HTMLButtonElement>>, any, {
    size: "small";
}, "size">;
