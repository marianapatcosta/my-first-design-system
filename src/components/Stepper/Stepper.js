var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useState, } from 'react';
import { ORIENTATION } from '../../constants';
import { Step } from '../Step/Step';
import { DoubleChevron } from '../../assets/icons';
import { StyledStepper, StyledStepperContainer, StyledStepperIndicator, StyledStepperIndicatorItem, StyledStepperProgressBar, StyledStepperContent, StyledStepperFooter, StyledStepperButton, StyledStepperButtonRight, } from './Stepper.style';
export var Stepper = function (_a) {
    var _b;
    var orientation = _a.orientation, stepsMetadata = _a.stepsMetadata, buttonLabel = _a.buttonLabel, onSubmit = _a.onSubmit, props = __rest(_a, ["orientation", "stepsMetadata", "buttonLabel", "onSubmit"]);
    var _c = useState(0), activeStepIndex = _c[0], setActiveStepIndex = _c[1];
    var progressPerStep = 100 / (stepsMetadata.length - 1);
    var isVertical = orientation === ORIENTATION.VERTICAL;
    var handleNextStepClick = function () {
        if (activeStepIndex === stepsMetadata.length - 1)
            return;
        setActiveStepIndex(function (prevActiveStepIndex) { return prevActiveStepIndex + 1; });
    };
    var handlePreviousStepClick = function () {
        if (activeStepIndex === 0)
            return;
        setActiveStepIndex(function (prevActiveStepIndex) { return prevActiveStepIndex - 1; });
    };
    var getProgressBarPercentage = function () {
        return activeStepIndex * progressPerStep;
    };
    return (_jsxs(StyledStepper, __assign({ "data-steps": stepsMetadata.length, isVertical: isVertical }, props, { children: [_jsxs(StyledStepperIndicator, __assign({ role: 'list', "aria-label": 'progress', isVertical: isVertical }, { children: [_jsx(StyledStepperProgressBar, { value: getProgressBarPercentage(), max: '100', isVertical: isVertical }, void 0),
                    stepsMetadata.map(function (step, index) { return (_jsx(StyledStepperIndicatorItem, __assign({ "aria-current": 'step', isFilled: index <= activeStepIndex }, { children: index + 1 }), "step-indicator-" + (index + Math.random()))); })] }), void 0),
            _jsxs(StyledStepperContainer, __assign({ isVertical: isVertical }, { children: [_jsx(StyledStepperContent, { children: stepsMetadata.map(function (_a, index) {
                            var content = _a.content;
                            return (_jsx(Step, { content: content, isActive: index === activeStepIndex }, "step-" + (index + Math.random())));
                        }) }, void 0),
                    _jsxs(StyledStepperFooter, { children: [activeStepIndex > 0 && (_jsx(StyledStepperButton, { "data-testid": 'prev-button', "aria-label": 'previous button', icon: DoubleChevron, onClick: handlePreviousStepClick }, void 0)),
                            _jsx(StyledStepperButtonRight, { "data-testid": 'next-button', "aria-label": activeStepIndex === stepsMetadata.length - 1
                                    ? buttonLabel + " button"
                                    : 'next button', label: activeStepIndex === stepsMetadata.length - 1 ? buttonLabel : '', icon: "" + (activeStepIndex !== stepsMetadata.length - 1 ? DoubleChevron : ''), disabled: !((_b = stepsMetadata[activeStepIndex].isValid) !== null && _b !== void 0 ? _b : true), onClick: activeStepIndex === stepsMetadata.length - 1
                                    ? onSubmit
                                    : handleNextStepClick }, void 0)] }, void 0)] }), void 0)] }), void 0));
};
Stepper.defaultProps = {
    id: '',
    className: '',
    style: {},
    orientation: ORIENTATION.LANDSCAPE,
    stepsMetadata: [],
    buttonLabel: 'submit',
    onSubmit: function () { return null; },
};
