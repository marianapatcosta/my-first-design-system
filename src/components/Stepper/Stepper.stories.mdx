import { Meta } from '@storybook/addon-docs/blocks';
import { Stepper } from './Stepper';

<Meta title="Design System|Stepper" component={Stepper} />

# Stepper

Stepper component

## Usage

Use stepper when makes sense to subdivide some content (e.g. a form) in smaller parts, which
should be rendered in Step component.

- Landscape Stepper:

  <Story name='Landscape'>
    <Stepper
      stepsMetadata={[
        { content: <p>Step 1 content</p> },
        { content: <p>Step 2 content</p> },
        { content: <p>Step 3 content</p> },
      ]}
      onSubmit={() => console.log('submitted')}
    />
  </Story>

- Portrait Stepper:
  <Story name='Portrait'>
    <Stepper
      orientation='portrait'
      stepsMetadata={[
        { content: <p>Step 1 content</p> },
        { content: <p>Step 2 content</p> },
        { content: <p>Step 3 content</p> },
      ]}
      onSubmit={() => console.log('submitted')}
    />
  </Story>

## Other important notes

not applicable.
