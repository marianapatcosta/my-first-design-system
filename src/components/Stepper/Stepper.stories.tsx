import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { Stepper, StepperProps } from './Stepper';

export default {
  title: 'Stepper',
  component: Stepper,
  args: {
    stepsMetadata: [
      {
        content: (
          <p style={{ textAlign: 'center', margin: '100px auto' }}>
            Step 1 content
          </p>
        ),
      },
      {
        content: (
          <p style={{ textAlign: 'center', margin: '100px auto' }}>
            Step 2 content
          </p>
        ),
      },
      {
        content: (
          <p style={{ textAlign: 'center', margin: '100px auto' }}>
            Step 3 content
          </p>
        ),
      },
    ],
    onSubmit: () => console.log('submitted'),
    style: { backgroundColor: 'gray' },
  },
} as Meta;

const Template: Story<StepperProps> = (args) => <Stepper {...args} />;

export const Landscape = Template.bind({});
Landscape.args = {};

export const Portrait = Template.bind({});
Portrait.args = { orientation: 'portrait' };
