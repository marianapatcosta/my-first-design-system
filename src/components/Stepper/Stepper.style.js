var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import styled from 'styled-components';
import { baseStyles, colors } from '../../styles';
import { Button } from '../Button/Button';
export var StyledStepper = styled.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", ";\n  display: flex;\n  flex-direction: column;\n  width: auto;\n  margin: 2rem auto;\n  padding: 1.5rem;\n  background-color: ", ";\n\n    ", ";\n"], ["\n  ", ";\n  display: flex;\n  flex-direction: column;\n  width: auto;\n  margin: 2rem auto;\n  padding: 1.5rem;\n  background-color: ", ";\n\n    ",
    ";\n"])), baseStyles, colors.primary, function (_a) {
    var isVertical = _a.isVertical;
    return isVertical &&
        "display: flex;\n  flex-direction: row;\n  justify-content: space-between;";
});
export var StyledStepperIndicator = styled.ul(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  display: flex;\n  align-items: center;\n  justify-content: space-between;\n  position: relative;\n  list-style: none;\n  background-color: inherit;\n\n  ", "\n"], ["\n  display: flex;\n  align-items: center;\n  justify-content: space-between;\n  position: relative;\n  list-style: none;\n  background-color: inherit;\n\n  ",
    "\n"])), function (_a) {
    var isVertical = _a.isVertical;
    return isVertical &&
        "  flex-direction: column;\n    margin: 2rem;\n    margin-bottom: 0;\n    ";
});
export var StyledStepperIndicatorItem = styled.li(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  width: 2.5rem;\n  height: 2.5rem;\n  background-color: inherit;\n  border: 0.125rem solid ", ";\n  border-radius: 50%;\n  -webkit-border-radius: 50%;\n  -moz-border-radius: 50%;\n  -ms-border-radius: 50%;\n  -o-border-radius: 50%;\n  z-index: 1;\n  transition: background-color 0.5s ease-in;\n  -webkit-transition: background-color 0.5s ease-in;\n  -moz-transition: background-color 0.5s ease-in;\n  -ms-transition: background-color 0.5s ease-in;\n  -o-transition: background-color 0.5s ease-in;\n\n  ", "\n"], ["\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  width: 2.5rem;\n  height: 2.5rem;\n  background-color: inherit;\n  border: 0.125rem solid ", ";\n  border-radius: 50%;\n  -webkit-border-radius: 50%;\n  -moz-border-radius: 50%;\n  -ms-border-radius: 50%;\n  -o-border-radius: 50%;\n  z-index: 1;\n  transition: background-color 0.5s ease-in;\n  -webkit-transition: background-color 0.5s ease-in;\n  -moz-transition: background-color 0.5s ease-in;\n  -ms-transition: background-color 0.5s ease-in;\n  -o-transition: background-color 0.5s ease-in;\n\n  ",
    "\n"])), colors.secondary, function (_a) {
    var isFilled = _a.isFilled;
    return isFilled &&
        " background-color: " + colors.secondary + ";  transition: all 0.5s ease-in;\n  -webkit-transition: all 0.5s ease-in;\n  -moz-transition: all 0.5s ease-in;\n  -ms-transition: background-color 0.5s ease-in;\n  -o-transition: background-color 0.5s ease-in;\n}";
});
export var StyledStepperProgressBar = styled.progress(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  position: absolute;\n  left: 0;\n  height: 0.125rem;\n  width: 100%;\n\n  ", "\n\n  &::-webkit-progress-bar {\n    background-color: ", ";\n  }\n\n  &::-webkit-progress-value {\n    background-color: ", ";\n    transition: width 0.3s ease-in, height 0.3s ease-in;\n    -webkit-transition: width 0.3s ease-in, height 0.3s ease-in;\n    -moz-transition: width 0.3s ease-in, height 0.3s ease-in;\n    -ms-transition: width 0.3s ease-in, height 0.3s ease-in;\n    -o-transition: width 0.3s ease-in, height 0.3s ease-in;\n  }\n\n  &[value] {\n    appearance: none;\n    border: 0;\n    color: ", ";\n    transition: width 0.3s ease-in, height 0.3s ease-in;\n    -webkit-transition: width 0.3s ease-in, height 0.3s ease-in;\n    -moz-transition: width 0.3s ease-in, height 0.3s ease-in;\n    -ms-transition: width 0.3s ease-in, height 0.3s ease-in;\n    -o-transition: width 0.3s ease-in, height 0.3s ease-in;\n  }\n"], ["\n  position: absolute;\n  left: 0;\n  height: 0.125rem;\n  width: 100%;\n\n  ",
    "\n\n  &::-webkit-progress-bar {\n    background-color: ", ";\n  }\n\n  &::-webkit-progress-value {\n    background-color: ", ";\n    transition: width 0.3s ease-in, height 0.3s ease-in;\n    -webkit-transition: width 0.3s ease-in, height 0.3s ease-in;\n    -moz-transition: width 0.3s ease-in, height 0.3s ease-in;\n    -ms-transition: width 0.3s ease-in, height 0.3s ease-in;\n    -o-transition: width 0.3s ease-in, height 0.3s ease-in;\n  }\n\n  &[value] {\n    appearance: none;\n    border: 0;\n    color: ", ";\n    transition: width 0.3s ease-in, height 0.3s ease-in;\n    -webkit-transition: width 0.3s ease-in, height 0.3s ease-in;\n    -moz-transition: width 0.3s ease-in, height 0.3s ease-in;\n    -ms-transition: width 0.3s ease-in, height 0.3s ease-in;\n    -o-transition: width 0.3s ease-in, height 0.3s ease-in;\n  }\n"])), function (_a) {
    var isVertical = _a.isVertical;
    return isVertical &&
        "left: revert;\n    bottom: 0;\n    top: 0;\n    margin: auto 0;\n    width: revert;\n    transform: rotate(90deg);";
}, colors.white, colors.secondary, colors.secondary);
export var StyledStepperContainer = styled.div(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  ", "\n"], ["\n  ", "\n"])), function (_a) {
    var isVertical = _a.isVertical;
    return isVertical && " width: 100%;";
});
export var StyledStepperContent = styled.div(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  margin: 1rem auto;\n  width: 100%;\n"], ["\n  margin: 1rem auto;\n  width: 100%;\n"])));
export var StyledStepperFooter = styled.footer(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n  display: flex;\n  align-items: center;\n"], ["\n  display: flex;\n  align-items: center;\n"])));
export var StyledStepperButton = styled(Button).attrs({ size: 'small' })(templateObject_8 || (templateObject_8 = __makeTemplateObject(["\n  margin-right: auto;\n  flex-direction: column;\n  img {\n    ", ";\n    padding: 0.15rem;\n    transform: rotate(90deg);\n  }\n"], ["\n  margin-right: auto;\n  flex-direction: column;\n  img {\n    ", ";\n    padding: 0.15rem;\n    transform: rotate(90deg);\n  }\n"])), colors.icon);
export var StyledStepperButtonRight = styled(StyledStepperButton)(templateObject_9 || (templateObject_9 = __makeTemplateObject(["\n  margin-right: 0;\n  margin-left: auto;\n  img {\n    transform: rotate(-90deg);\n  }\n"], ["\n  margin-right: 0;\n  margin-left: auto;\n  img {\n    transform: rotate(-90deg);\n  }\n"])));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8, templateObject_9;
