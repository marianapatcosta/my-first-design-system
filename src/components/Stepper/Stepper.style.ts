import styled from 'styled-components';
import { baseStyles, colors } from '../../styles';
import { Button } from '../Button/Button';

export const StyledStepper = styled.div<{ isVertical?: boolean }>`
  ${baseStyles};
  display: flex;
  flex-direction: column;
  width: auto;
  margin: 2rem auto;
  padding: 1.5rem;
  background-color: ${colors.primary};

  ${({ isVertical }) =>
    isVertical &&
    `display: flex;
  flex-direction: row;
  justify-content: space-between;`};
`;

export const StyledStepperIndicator = styled.ul<{ isVertical?: boolean }>`
  display: flex;
  align-items: center;
  justify-content: space-between;
  position: relative;
  list-style: none;
  background-color: inherit;

  ${({ isVertical }) =>
    isVertical &&
    `  flex-direction: column;
    margin: 2rem;
    margin-bottom: 0;
    `}
`;

export const StyledStepperIndicatorItem = styled.li<{ isFilled?: boolean }>`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 2.5rem;
  height: 2.5rem;
  background-color: inherit;
  border: 0.125rem solid ${colors.secondary};
  border-radius: 50%;
  -webkit-border-radius: 50%;
  -moz-border-radius: 50%;
  -ms-border-radius: 50%;
  -o-border-radius: 50%;
  z-index: 1;
  transition: background-color 0.5s ease-in;
  -webkit-transition: background-color 0.5s ease-in;
  -moz-transition: background-color 0.5s ease-in;
  -ms-transition: background-color 0.5s ease-in;
  -o-transition: background-color 0.5s ease-in;

  ${({ isFilled }) =>
    isFilled &&
    ` background-color: ${colors.secondary};  transition: all 0.5s ease-in;
  -webkit-transition: all 0.5s ease-in;
  -moz-transition: all 0.5s ease-in;
  -ms-transition: background-color 0.5s ease-in;
  -o-transition: background-color 0.5s ease-in;
}`}
`;

export const StyledStepperProgressBar = styled.progress<{
  isVertical?: boolean;
}>`
  position: absolute;
  left: 0;
  height: 0.125rem;
  width: 100%;

  ${({ isVertical }) =>
    isVertical &&
    `left: revert;
    bottom: 0;
    top: 0;
    margin: auto 0;
    width: revert;
    transform: rotate(90deg);`}

  &::-webkit-progress-bar {
    background-color: ${colors.white};
  }

  &::-webkit-progress-value {
    background-color: ${colors.secondary};
    transition: width 0.3s ease-in, height 0.3s ease-in;
    -webkit-transition: width 0.3s ease-in, height 0.3s ease-in;
    -moz-transition: width 0.3s ease-in, height 0.3s ease-in;
    -ms-transition: width 0.3s ease-in, height 0.3s ease-in;
    -o-transition: width 0.3s ease-in, height 0.3s ease-in;
  }

  &[value] {
    appearance: none;
    border: 0;
    color: ${colors.secondary};
    transition: width 0.3s ease-in, height 0.3s ease-in;
    -webkit-transition: width 0.3s ease-in, height 0.3s ease-in;
    -moz-transition: width 0.3s ease-in, height 0.3s ease-in;
    -ms-transition: width 0.3s ease-in, height 0.3s ease-in;
    -o-transition: width 0.3s ease-in, height 0.3s ease-in;
  }
`;

export const StyledStepperContainer = styled.div<{ isVertical?: boolean }>`
  ${({ isVertical }) => isVertical && ` width: 100%;`}
`;

export const StyledStepperContent = styled.div<{ isFilled?: boolean }>`
  margin: 1rem auto;
  width: 100%;
`;

export const StyledStepperFooter = styled.footer<{ isFilled?: boolean }>`
  display: flex;
  align-items: center;
`;

export const StyledStepperButton = styled(Button).attrs({ size: 'small' })`
  margin-right: auto;
  flex-direction: column;
  img {
    filter: ${colors.icon};
    padding: 0.15rem;
    transform: rotate(90deg);
  }
`;

export const StyledStepperButtonRight = styled(StyledStepperButton)`
  margin-right: 0;
  margin-left: auto;
  img {
    transform: rotate(-90deg);
  }
`;
