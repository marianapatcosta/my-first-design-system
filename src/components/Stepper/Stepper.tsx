import React, {
  CSSProperties,
  MouseEventHandler,
  ReactNode,
  useState,
} from 'react';
import { ORIENTATIONS } from '../../constants';
import { Step } from '../Step/Step';
import { DoubleChevron } from '../../assets/icons';
import {
  StyledStepper,
  StyledStepperContainer,
  StyledStepperIndicator,
  StyledStepperIndicatorItem,
  StyledStepperProgressBar,
  StyledStepperContent,
  StyledStepperFooter,
  StyledStepperButton,
  StyledStepperButtonRight,
} from './Stepper.style';

type orientationTypes = ORIENTATIONS.PORTRAIT | ORIENTATIONS.LANDSCAPE;

export interface StepperProps {
  /**
   * element id
   */
  id?: string;
  /**
   * additional css class
   */
  className?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
  /**
   * additional style
   */
  orientation?: orientationTypes;
  /**
   * Array of objects defining the metadata for each stepper's steps
   */
  stepsMetadata: Array<{
    content: ReactNode;
    isValid?: boolean;
  }>;
  /**
   * Click handler to call on finish the fulfill/visualization of all steps
   */
  onSubmit: MouseEventHandler<HTMLButtonElement>;
}

export const Stepper: React.FC<StepperProps> = ({
  orientation,
  stepsMetadata,
  onSubmit,
  ...props
}) => {
  const [activeStepIndex, setActiveStepIndex] = useState<number>(0);
  const progressPerStep = 100 / (stepsMetadata.length - 1);
  const isPortrait: boolean = orientation === ORIENTATIONS.PORTRAIT;

  const handleNextStepClick = (): void => {
    if (activeStepIndex === stepsMetadata.length - 1) return;
    setActiveStepIndex((prevActiveStepIndex) => prevActiveStepIndex + 1);
  };

  const handlePreviousStepClick = (): void => {
    if (activeStepIndex === 0) return;
    setActiveStepIndex((prevActiveStepIndex) => prevActiveStepIndex - 1);
  };

  const getProgressBarPercentage = (): number =>
    activeStepIndex * progressPerStep;
  return (
    <StyledStepper
      data-steps={stepsMetadata.length}
      isPortrait={isPortrait}
      {...props}
    >
      <StyledStepperIndicator
        role='group'
        aria-label='progress'
        isPortrait={isPortrait}
      >
        <StyledStepperProgressBar
          value={getProgressBarPercentage()}
          max='100'
          isPortrait={isPortrait}
        />
        {stepsMetadata.map((step: string | ReactNode, index: number) => (
          <StyledStepperIndicatorItem
            key={`step-indicator-${index + Math.random()}`}
            aria-current={'step'}
            isFilled={index <= activeStepIndex}
          >
            {index + 1}
          </StyledStepperIndicatorItem>
        ))}
      </StyledStepperIndicator>
      <StyledStepperContainer isPortrait={isPortrait}>
        <StyledStepperContent>
          {stepsMetadata.map(({ content }, index) => (
            <Step
              key={`step-${index + Math.random()}`}
              content={content}
              isActive={index === activeStepIndex}
            />
          ))}
        </StyledStepperContent>
        <StyledStepperFooter>
          {activeStepIndex > 0 && (
            <StyledStepperButton
              data-testid='prev-button'
              icon={DoubleChevron}
              onClick={handlePreviousStepClick}
            />
          )}
          <StyledStepperButtonRight
            data-testid='next-button'
            label={activeStepIndex === stepsMetadata.length - 1 ? 'submit' : ''}
            icon={`${
              activeStepIndex !== stepsMetadata.length - 1 ? DoubleChevron : ''
            }`}
            disabled={!(stepsMetadata[activeStepIndex].isValid ?? true)}
            onClick={
              activeStepIndex === stepsMetadata.length - 1
                ? onSubmit
                : handleNextStepClick
            }
          />
        </StyledStepperFooter>
      </StyledStepperContainer>
    </StyledStepper>
  );
};

Stepper.defaultProps = {
  id: '',
  className: '',
  style: {},
  orientation: ORIENTATIONS.LANDSCAPE,
  stepsMetadata: [],
  onSubmit: () => null,
};
