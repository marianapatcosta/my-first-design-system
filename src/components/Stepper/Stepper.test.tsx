import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { ORIENTATION } from '../../constants';
import { Stepper } from './Stepper';

describe('Stepper', () => {
  const defaultProps = {
    stepsMetadata: [
      {
        content: <p>Step 1 content</p>,
      },
      {
        content: <p>Step 2 content</p>,
      },
      {
        content: <p>Step 3 content</p>,
      },
    ],
    onSubmit: () => null,
  };

  test('should render without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<Stepper {...defaultProps} />);

    expect(firstChild).toMatchSnapshot();
  });

  test('should render step 1 content by default', () => {
    const { getByText } = render(<Stepper {...defaultProps} />);

    getByText(/Step 1/i);
    !getByText(/Step 2/i);
    !getByText(/Step 3/i);
  });

  test('should render `next` button but not `prev` button when rendering', () => {
    const { queryByTestId } = render(<Stepper {...defaultProps} />);

    expect(queryByTestId('next-button')).toBeInTheDocument();
    expect(queryByTestId('prev-button')).not.toBeInTheDocument();
  });

  test('should render step 2 content on `next` button click', () => {
    const { getByText, getByTestId } = render(<Stepper {...defaultProps} />);

    const nextButton = getByTestId('next-button');
    nextButton && fireEvent.click(nextButton);

    getByText(/Step 2/i);
    !getByText(/Step 1/i);
    !getByText(/Step 3/i);
  });

  test('should render step 1 content after `next` button click and `prev`button click', () => {
    const { getByText, getByTestId } = render(<Stepper {...defaultProps} />);

    const nextButton = getByTestId('next-button');
    nextButton && fireEvent.click(nextButton);
    const prevButton = getByTestId('prev-button');
    prevButton && fireEvent.click(prevButton);

    getByText(/Step 1/i);
    !getByText(/Step 2/i);
    !getByText(/Step 3/i);
  });

  test('should call `onSubmit` on `submit` click', () => {
    const onSubmitMock = jest.fn();
    const { getByTestId } = render(
      <Stepper {...defaultProps} onSubmit={onSubmitMock} />
    );

    const nextButton = getByTestId('next-button');
    nextButton && fireEvent.click(nextButton);
    nextButton && fireEvent.click(nextButton);
    nextButton && fireEvent.click(nextButton);

    expect(onSubmitMock).toBeCalled();
  });

  test('should have default styles if orientation= landscape', () => {
    const {
      container: { firstChild },
      getByLabelText,
    } = render(<Stepper {...defaultProps} />);

    const stepperIndicator = getByLabelText('progress');
    const stepperStyle =
      firstChild && window.getComputedStyle(firstChild as Element);
    const stepperIndicatorStyle = window.getComputedStyle(stepperIndicator);

    expect(stepperStyle?.flexDirection).toBe('column');
    expect(stepperIndicatorStyle.flexDirection).toBe('');
  });

  test('should have `isVertical` styles if orientation = vertical', () => {
    const {
      container: { firstChild },
      getByLabelText,
    } = render(
      <Stepper {...defaultProps} orientation={ORIENTATION.VERTICAL} />
    );

    const stepperIndicator = getByLabelText('progress');

    const stepperStyle =
      firstChild && window.getComputedStyle(firstChild as Element);
    const stepperIndicatorStyle = window.getComputedStyle(stepperIndicator);

    stepperStyle && expect(stepperStyle.flexDirection).toBe('row');
    expect(stepperIndicatorStyle.flexDirection).toBe('column');
  });
});
