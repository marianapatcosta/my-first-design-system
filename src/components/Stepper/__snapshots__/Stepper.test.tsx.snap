// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper should render without errors and matches snapshot 1`] = `
<div
  class="sc-eCssSg kPWSdA"
  data-steps="3"
  id=""
>
  <ul
    aria-label="progress"
    class="sc-jSgupP bFkA-dv"
    role="group"
  >
    <progress
      class="sc-iBPRYJ hsnDam"
      max="100"
      value="0"
    />
    <li
      aria-current="step"
      class="sc-gKsewC iQAMcE"
    >
      1
    </li>
    <li
      aria-current="step"
      class="sc-gKsewC FsmaJ"
    >
      2
    </li>
    <li
      aria-current="step"
      class="sc-gKsewC FsmaJ"
    >
      3
    </li>
  </ul>
  <div
    class="sc-fubCfw kVPOkw"
  >
    <div
      class="sc-pFZIQ fiTKve"
    >
      <div
        class="sc-bdfBwQ kcDtUs"
        id=""
      >
        <p>
          Step 1 content
        </p>
      </div>
      <div
        class="sc-bdfBwQ gmDwpK"
        id=""
      >
        <p>
          Step 2 content
        </p>
      </div>
      <div
        class="sc-bdfBwQ gmDwpK"
        id=""
      >
        <p>
          Step 3 content
        </p>
      </div>
    </div>
    <footer
      class="sc-jrAGrp bykmwW"
    >
      <button
        class="sc-gsTCUz fOjZNc sc-kEjbxe sc-iqHYGH loVkLi ijhVeU"
        data-testid="next-button"
        id=""
        type="button"
      >
        <img
          alt=""
          class="sc-hKgILt VuIIP"
          data-testid="button-icon"
          label=""
          src="double-chevron.svg"
        />
        
      </button>
    </footer>
  </div>
</div>
`;
